controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -949,6 +993,28 @@ func (r *BareMetalHostReconciler) getHostStatusFromAnnotation(host *metal3v1alph
 	return objStatus, nil
 }
 
+func unmarshalHardwareDetailsAnnotation(content []byte) (*metal3v1alpha1.HardwareDetails, error) {
+	objHardwareDetails := &metal3v1alpha1.HardwareDetails{}
+	if err := json.Unmarshal(content, objHardwareDetails); err != nil {
+		return nil, errors.Wrap(err, "Failed to fetch HardwareDetails from annotation")
+	}
+	return objHardwareDetails, nil
+}
+
+// extract host from HardwareDetails annotation
+func (r *BareMetalHostReconciler) getHardwareDetailsFromAnnotation(host *metal3v1alpha1.BareMetalHost) (*metal3v1alpha1.HardwareDetails, error) {
+	annotations := host.GetAnnotations()
+	if annotations[hardwareDetailsAnnotation] == "" {
+		return nil, nil
+	}
+	content := []byte(annotations[hardwareDetailsAnnotation])
+	objHardwareDetails, err := unmarshalHardwareDetailsAnnotation(content)
```

Comment:
We can just return unmarshalHardwareDetailsAnnotation(content) here. unmarshalHardwareDetailsAnnotation() is already returning the right values whether there is an error or not (nil, err and objHardwareDetails, nil).

Reply:
Yeah good point, it can probably be simplified into a single function, please check the latest version - thanks for the review!